group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java

```diff
@@ -0,0 +1,240 @@
+                    RemainingAssignmentsForMember<String, Integer> newPair = new RemainingAssignmentsForMember<>(memberId, remaining);
+                    potentiallyUnfilledMembers.add(newPair);
+                }
+            }
+
+            // Step 3
+            // If remaining > 0 after increasing the required quota due to the extra partition, add potentially unfilled member to the unfilled members list.
+            for (RemainingAssignmentsForMember<String, Integer> pair : potentiallyUnfilledMembers) {
+                String memberId = pair.memberId();
+                Integer remaining = pair.remaining();
+                if (numMembersWithExtraPartition > 0) {
+                    remaining++;
+                    numMembersWithExtraPartition--;
+                }
+                if (remaining > 0) {
+                    RemainingAssignmentsForMember<String, Integer> newPair = new RemainingAssignmentsForMember<>(memberId, remaining);
+                    unfilledMembersPerTopic.computeIfAbsent(topicId, k -> new ArrayList<>()).add(newPair);
+                }
+            }
+        });
+
+        // Step 4
+        // Find the difference between the total partitions per topic and the already assigned sticky partitions for the topic to get unassigned partitions.
+        Map<Uuid, List<Integer>> unassignedPartitionsPerTopic = getUnassignedPartitionsPerTopic(assignmentSpec, assignedStickyPartitionsPerTopic);
+
+        // Step 5
+        // Iterate through the unfilled members list and assign the remaining number of partitions from the unassignedPartitions list.
+        unfilledMembersPerTopic.forEach((topicId, unfilledMembersForTopic) -> {
+            int unassignedPartitionsListStartPointer = 0;
+            for (RemainingAssignmentsForMember<String, Integer> remainingAssignmentsForMember : unfilledMembersForTopic) {
+                String memberId = remainingAssignmentsForMember.memberId();
+                int remaining = remainingAssignmentsForMember.remaining();
+                List<Integer> partitionsToAssign = unassignedPartitionsPerTopic.get(topicId).subList(unassignedPartitionsListStartPointer, unassignedPartitionsListStartPointer + remaining);
+                unassignedPartitionsListStartPointer += remaining;
+                membersWithNewAssignmentPerTopic.computeIfAbsent(memberId, k -> new HashMap<>()).computeIfAbsent(topicId, k -> new HashSet<>()).addAll(partitionsToAssign);
+            }
+        });
+
+        // Consolidate the maps into MemberAssignment and then finally map each member to a MemberAssignment.
+        Map<String, MemberAssignment> membersWithNewAssignment = new HashMap<>();
+
+        membersWithNewAssignmentPerTopic.forEach((memberId, assignmentPerTopic) -> membersWithNewAssignment.computeIfAbsent(memberId, k -> new MemberAssignment(assignmentPerTopic)));
```

Comment:
nit: I have noticed that you use `computeIfAbsent` in a few places where `put` would just work.

Reply:
I removed most of them and converted them to put before, I'll check again

Reply:
Re-checked and computeIfAbsent is the best safe way to do it